atomicapp/index.py

```diff
@@ -0,0 +1,203 @@
+            providers = ""
+            for provider in entry["providers"]:
+                providers = "%s,%s" % (providers, provider[0].capitalize())
+
+            # Remove the first element, add brackets
+            providers = "{%s}" % providers[1:]
+
+            # Retrieve the entry information
+            id = entry.get('id') or ""
+            version = entry['metadata'].get('appversion') or ""
+            location = entry['metadata'].get('location') or INDEX_DEFAULT_IMAGE_LOCATION
+
+            # Print out the row
+            print(index_format.format(
+                id,
+                version,
+                providers,
+                location))
+
+    def update(self, index_image=INDEX_IMAGE):
+        """
+        Fetch the latest index image and update the file based upon
+        the INDEX_IMAGE attribute. By default, this should pull the
+        'official' Nulecule index.
+        """
+
+        logger.info("Updating the index list")
+        logger.info("Pulling latest index image...")
+        self._fetch_index_container()
+        logger.info("Index updated")
+
+    # TODO: Error out if the locaiton does not have a Nulecule file / dir
+    def generate(self, location, output_location=INDEX_GEN_DEFAULT_OUTPUT_LOC):
+        """
+        Generate an index.yaml with a provided directory location
+        """
+        logger.info("Generating index.yaml from %s" % location)
+        self.index = deepcopy(self.index_template)
+
+        if not os.path.isdir(location):
+            raise Exception("Location must be a directory")
+
+        for f in os.listdir(location):
+            nulecule_dir = os.path.join(location, f)
+            if f.startswith("."):
+                continue
+            if os.path.isdir(nulecule_dir):
+                try:
+                    index_info = self._nulecule_get_info(nulecule_dir)
+                except NuleculeException as e:
+                    logger.warning("SKIPPING %s. %s" %
+                                   (nulecule_dir, e))
+                    continue
+                index_info["path"] = f
+                self.index["nulecules"].append(index_info)
+
+        if len(index_info) > 0:
+            anymarkup.serialize_file(self.index, output_location, format="yaml")
+        logger.info("index.yaml generated")
+
+    def _fetch_index_container(self, index_image=INDEX_IMAGE):
+        """
+        Fetch the index container
+        """
+        # Create the ".atomicapp" dir if it does not exist
+        if not os.path.exists(os.path.dirname(self.index_location)):
+            try:
+                os.makedirs(os.path.dirname(self.index_location))
+            except OSError as exc:  # Guard against race condition
+                if exc.errno != errno.EEXIST:
+                    raise
+
+        dh = DockerHandler()
+        dh.pull(index_image)
+        dh.extract_files(index_image, "/" + INDEX_NAME, self.index_location)
```

Comment:
hmm ok - so running this in the "in container" use case doesn't work. It doesn't copy files to ~/.atomicapp/ - it copies them to that dir in the container but not outside the container.

one thing to consider is do we need to store things on the machine or could we just leave the index in the container and cp it out each time the user runs the command.